### ISSUE TITLE
Variables named `name` are replaced with the result of calling their `.toString` method

### ISSUE THREAD

**Issue by ds300** [jsbin/jsbin#2849]
related to #1070 and ds300/derivablejs#47

Demo: http://jsbin.com/migaxacuwe/edit?js,console

Screenshot for posterity:

![image](https://cloud.githubusercontent.com/assets/1242537/17768037/9b715fd2-6532-11e6-9ada-f3264d303733.png)

**Comment by remy** [comment]
This is actually a quirk to do with the `name` variable.

`name` already exists in scope, though you're able (I think) to redefine it using `const` (this is where my understanding gets a bit hazy). But to see the exact same behaviour, try running your code (once converted to ES5) on a page like google.com (in the browser devtools), you'll see the same result.

```js
console.clear();
var name = {
    toString: function () {
        console.log("uh oh, I called name.toString()");
    },
};
alert(JSON.stringify(name))
console.log("name === 'undefined': " + (name === 'undefined'));
```

I also wonder if the TypeScript transpile of `const` is using `var` *instead* of `const` which is masking the issue. If you run the above code, but change `var` for `const` it does actually work.

Never the less, this isn't actually a JS Bin issue, but certainly an interesting edge case.

Bottom line: don't use `name`…which is a bit random really (but I've seen other people run into this before too).

**Comment by ds300** [comment]
Thanks for the detailed response. Just one more of JS's many many esoteric delights 🌈
